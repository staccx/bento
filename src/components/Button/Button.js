import PropTypes from "prop-types"
import React from "react"
import styled from "styled-components"
import {
  color,
  borderRadius,
  spacing,
  targetSize,
  font,
  fontFamily,
  fontWeight,
  applyVariants,
  commonPropTypes
} from "../../theming"
import themeProps from "./Button.themeProps"
import { componentCreateFactory } from "../../theming/utils/createVariantsFunctionFactory"

const ButtonComp = styled.button`
  background-color: ${color.primary};
  color: ${color.white};
  border-width: 0;
  border-radius: ${borderRadius};
  padding: ${spacing.tiny} ${spacing.medium};
  cursor: pointer;
  min-height: ${targetSize.normal};
  text-decoration: none;
  text-align: center;
  display: inline-block;
  font-size: ${font.base};
  font-family: ${fontFamily.body};
  line-height: 32px;
  font-weight: ${fontWeight.bold};
  white-space: nowrap;
  justify-content: center;
  align-items: center;
  -webkit-font-smoothing: antialiased;
  -moz-osx-font-smoothing: grayscale;
  transition: background 0.2s ease;

  &:hover,
  &:focus {
    outline: none;
    background-color: ${color.primary};
    color: ${color.white};
  }

  ${applyVariants(themeProps.button)};
`
/**
 * Button component
 */
const Button = props => <ButtonComp {...props} />

const defaultProps = {
  children: "button",
  onClick: null
}

const propTypes = {
  children: commonPropTypes.children.isRequired,
  /**
   * Click callback. Lets you do stuff onClick
   * @export
   */
  onClick: PropTypes.func
}

Button.defaultProps = defaultProps
Button.propTypes = propTypes
Button.themeProps = themeProps
Button.createVariants = componentCreateFactory(Button)

export default Button
