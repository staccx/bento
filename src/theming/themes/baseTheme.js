import React from "react"
import Theme from "../Theme"
import reset from "../styles/cssResets"
import grids from "../../components/Layout/Layout/Layout.Grids"
import codeTheme from "../codeTheme"

const Logo = () => <div>ThemeComponent for BaseTheme</div>
Logo.displayName = "Logo for base theme"

export const theme = {
  name: "Base Theme",
  storybookName: "Base Theme",
  spacing: {
    huge: "96px",
    large: "48px",
    mediumPlus: "32px",
    medium: "24px",
    small: "12px",
    tiny: "6px",
    micro: "3px"
  },
  wrapper: {
    small: "420px",
    medium: "640px",
    large: "1024px",
    breakout: "640px"
  },
  color: {
    white: "#fff",
    bg: "#fff",
    subtleHover: "#F7F7FF",
    bgGray: "#f7f7ff",
    line: "#EBEBF2",
    gray: "#767676",
    wcag: "#777",
    disabled: "#BABABA",
    primary: "#00a190",
    secondary: "#F5A623",
    text: "#2b2b2b",
    negative: "#ea526f",
    positive: "#4A90E2",
    warning: "#c54059",
    black: "#000",
    green: "#00a190", // for toggle eks
    link: "#279AF1"
  },
  fontFamily: {
    heading: "sans-serif",
    body: "sans-serif"
  },
  font: {
    h1: "36px",
    h2: "26px",
    h3: "22px",
    h4: "18px",
    h5: "16px",
    h6: "14px",
    base: "14px",
    tiny: "12px",
    input: "16px",
    huge: "44px",
    lineHeight: "1.5"
  },
  fontWeight: {
    bold: "bold",
    normal: "normal",
    light: "lighter"
  },
  targetSize: {
    small: "36px",
    normal: "44px"
  },
  borderRadius: "2px",
  reset,
  grid: grids,
  logo: Logo
}

export const StyleguideTheme = new Theme(theme, codeTheme)

export default theme
