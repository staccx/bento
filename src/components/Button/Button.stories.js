import React from "react"
import Button from "./Button"

export default {
  title: "Components/Button",
  component: Button
}

export const text = () => <Button> Button </Button>
