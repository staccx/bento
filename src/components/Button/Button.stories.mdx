import {useState} from "react"
import { Meta, Story, Preview, Props } from '@storybook/addon-docs/blocks';
import { action }Â from "@storybook/addon-actions"
import Button from "./Button"

<Meta title="components/Button" component={Button} parameters={{component: Button.themeProps.button.name}} />

# Button
Bento Button Component

```javascript
import {Button} from "@staccx/bento"
```


## Standard Button
This is the standard button

<Preview>
  <Story name="Standard">
   <Button>Knapp</Button>
  </Story>
</Preview>


## Long String
<Preview>
  <Story name="Long String">
   <Button>This is a button with long text</Button>
  </Story>
</Preview>

## With useState
<Preview>
  <Story name="with counter">
    {() => {
      const [counter, setCounter] = useState(0);
      const label = `Testing: ${counter}`;
      return <Button onClick={() => setCounter(counter + 1)}>{label}</Button>;
    }}
  </Story>
</Preview>

### Source code

<details><summary>Button</summary>
<p>

```bash
import PropTypes from "prop-types"
import React from "react"
import styled from "styled-components"
import {
  color,
  borderRadius,
  spacing,
  targetSize,
  font,
  fontFamily,
  fontWeight,
  applyVariants,
  commonPropTypes
} from "../../theming"
import themeProps from "./Button.themeProps"
import { componentCreateFactory } from "../../theming/utils/createVariantsFunctionFactory"

const ButtonComp = styled.button`
  background-color: ${color.primary};
  color: ${color.white};
  border-width: 0;
  border-radius: ${borderRadius};
  padding: ${spacing.tiny} ${spacing.medium};
  cursor: pointer;
  min-height: ${targetSize.normal};
  text-decoration: none;
  text-align: center;
  display: inline-block;
  font-size: ${font.base};
  font-family: ${fontFamily.body};
  line-height: 32px;
  font-weight: ${fontWeight.bold};
  white-space: nowrap;
  justify-content: center;
  align-items: center;
  -webkit-font-smoothing: antialiased;
  -moz-osx-font-smoothing: grayscale;
  transition: background 0.2s ease;

  &:hover,
  &:focus {
    outline: none;
    background-color: ${color.primary};
    color: ${color.white};
  }

  ${applyVariants(themeProps.button)};
`
/**
 * Button component. Use to click stuff
 */
const Button = props => <ButtonComp {...props} />

const defaultProps = {
  children: "button",
  onClick: null
}

const propTypes = {
  children: commonPropTypes.children.isRequired,
  /**
   * Click callback. Lets you do stuff onClick
   * @export
   */
  onClick: PropTypes.func
}

Button.defaultProps = defaultProps
Button.propTypes = propTypes
Button.themeProps = themeProps
Button.createVariants = componentCreateFactory(Button)

export default Button

```

</p>
</details>


## Props for button
<Props of={Button} />
