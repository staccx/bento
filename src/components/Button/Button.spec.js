import React from "react"
import renderer from "react-test-renderer"
import { fireEvent, render, screen } from "@testing-library/react"
import { ThemeProvider } from "styled-components"
import baseTheme from "../../theming/themes/baseTheme"
import Button from "./Button"

const text = "Test"

describe("Button", () => {
  describe("Snapshots", () => {
    it("No props", () => {
      const tree = renderer
        .create(
          <ThemeProvider theme={baseTheme}>
            <Button>{text}</Button>
          </ThemeProvider>
        )
        .toJSON()
      expect(tree).toMatchSnapshot()
    })
    it("with onClick", () => {
      const tree = renderer
        .create(
          <ThemeProvider theme={baseTheme}>
            <Button onClick={() => console.log("Clicked")}>{text}</Button>
          </ThemeProvider>
        )
        .toJSON()
      expect(tree).toMatchSnapshot()
    })
    it("no children", () => {
      const tree = renderer
        .create(
          <ThemeProvider theme={baseTheme}>
            <Button onClick={() => console.log("Clicked")} />
          </ThemeProvider>
        )
        .toJSON()
      expect(tree).toMatchSnapshot()
    })
  })
  describe("Rendering", () => {
    it("Should click", () => {
      console.log = jest.fn()
      render(
        <ThemeProvider theme={baseTheme}>
          <Button onClick={() => console.log("click")}>{text}</Button>
        </ThemeProvider>
      )
      fireEvent.click(screen.getByText("Test"))
      expect(console.log.mock.calls[0][0]).toBe("click")
    })
    it("Should forward ref", () => {
      const ref = React.createRef()
      render(<Button ref={ref}>{text}</Button>)
      expect(ref.current).not.toBeNull()
    })
  })
})
