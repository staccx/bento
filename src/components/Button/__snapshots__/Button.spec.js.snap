// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button Snapshots No props 1`] = `
<button
  className="sc-bdfBwQ cXIOkm"
  onClick={null}
>
  Test
</button>
`;

exports[`Button Snapshots no children 1`] = `
<button
  className="sc-bdfBwQ cXIOkm"
  onClick={[Function]}
>
  button
</button>
`;

exports[`Button Snapshots with onClick 1`] = `
<button
  className="sc-bdfBwQ cXIOkm"
  onClick={[Function]}
>
  Test
</button>
`;
